src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java

```diff
@@ -453,32 +455,73 @@ private CompletableFuture<Response> checkFor407(ExchangeImpl<T> ex, Throwable t,
     // for the 100-Continue response
     private CompletableFuture<Response> expectContinue(ExchangeImpl<T> ex) {
         assert request.expectContinue();
+
+        long responseTimeout = 5;
+        if (request.timeout().isPresent() && request.timeout().get().getSeconds() < responseTimeout) {
+            responseTimeout = request.timeout().get().getSeconds();
+        }
+
         return ex.getResponseAsync(parentExecutor)
+                .orTimeout(responseTimeout, TimeUnit.SECONDS)
```

Comment:
Hello Darragh, would it be wise to check for the `responseTimeout` to be greater than `0`? I see that we set it to the request timeout value above and if the request timeout is say 500 milli seconds, then `responseTimeout` will end up being `0`.
`CompletableFuture.orTimeout()` doesn't specify how it deals with `0` or negative time value to the parameter of that method. So I think we might have to do some checks here to avoid passing `0`. Maybe we should just use nanos as a time unit?

Reply:
Interesting. If requestTimeout is 500ms I guess it means we will send the body immediately without waiting. Possibly the best course of action here could be to use millis as the granularity. Given the latency of HTTP then anything below 1ms is probably equivalent to 0 (no wait).

Reply:
Thats a good point, changing the timeunit to nanos is probably for the best.

I wonder in the case of timeout being equal to 0 what the best way to handle it is though? 
perhaps something like
```         
        long responseTimeout = TimeUnit.SECONDS.toNanos(5);
        if (request.timeout().isPresent()
                && request.timeout().get().getNano() > 0
                && request.timeout().get().getNano() < responseTimeout) {
            responseTimeout = request.timeout().get().getNano();
        }

Reply:
> I wonder in the case of timeout being equal to 0 what the best way to handle it is though?

The `HttpRequest.Builder.timeout()` doesn't allow for `0` (or negative) value to be specified for the timeout duration. So if the request timeout is indeed specified, then we can be sure that the nanos representation of it here will be greater than 0. We could just add an `assert` for that, after converting it to nanos (when the request timeout is present). Something like this untested code:

```java
long timeoutNanos = TimeUnit.SECONDS.toNanos(5);
if(request.timeout().isPresent() {
	final long rtNanos = request.timeout().get().getNano();
	assert rtNanos > 0 : "unexpected request timeout: " + rtNanos;
	timeoutNanos = Math.min(timeoutNanos, rt);
});

 
```

Reply:
I would just let 0 be 0: no need to split 1ms in 500_000 nanos :-)

Reply:
That sounds fine to me.